### Exercise: 
Machine Learning Finding Optimal Model and Hyperparameters
For digits dataset in sklearn.dataset, please try following classifiers and find out the one that gives best performance. Also find the optimal parameters for that classifier.

1. svm
2. RandomForestClassifier
3. LogisticRegression
4. GaussianNB
5. MultinomialNB
6. DecisionTreeClassifier

In [53]:
import pandas as pd
from sklearn.datasets import load_digits
import warnings
warnings.filterwarnings('ignore')

In [54]:
digit=load_digits()

In [55]:
dir(digit)

['DESCR', 'data', 'feature_names', 'frame', 'images', 'target', 'target_names']

In [56]:
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.linear_model import LogisticRegression
from sklearn.naive_bayes import GaussianNB
from sklearn.naive_bayes import MultinomialNB
from sklearn.tree import DecisionTreeClassifier

### GridSearchCV for Random Forest

In [57]:
from sklearn.model_selection import GridSearchCV

In [58]:
clf=GridSearchCV(RandomForestClassifier(),{
    'n_estimators':[5,10,15,100,150]
},return_train_score=False,cv=3
)

In [59]:
clf.fit(digit.data,digit.target)

GridSearchCV(cv=3, estimator=RandomForestClassifier(),
             param_grid={'n_estimators': [5, 10, 15, 100, 150]})

In [60]:
clf.cv_results_

{'mean_fit_time': array([0.0228556 , 0.05212561, 0.07285714, 0.4940877 , 0.74882762]),
 'std_fit_time': array([0.00677847, 0.00825978, 0.00221882, 0.00366722, 0.00647466]),
 'mean_score_time': array([0.        , 0.00791518, 0.01015615, 0.03078024, 0.04403877]),
 'std_score_time': array([0.00000000e+00, 1.86842275e-03, 3.64247621e-05, 3.54466363e-04,
        5.00609867e-03]),
 'param_n_estimators': masked_array(data=[5, 10, 15, 100, 150],
              mask=[False, False, False, False, False],
        fill_value='?',
             dtype=object),
 'params': [{'n_estimators': 5},
  {'n_estimators': 10},
  {'n_estimators': 15},
  {'n_estimators': 100},
  {'n_estimators': 150}],
 'split0_test_score': array([0.85141903, 0.89482471, 0.9115192 , 0.93489149, 0.93322204]),
 'split1_test_score': array([0.85642738, 0.91819699, 0.93656093, 0.95158598, 0.96160267]),
 'split2_test_score': array([0.83472454, 0.89983306, 0.9015025 , 0.92821369, 0.92654424]),
 'mean_test_score': array([0.84752365, 0.9042

In [61]:
df=pd.DataFrame(clf.cv_results_)
df

mean_fit_time  std_fit_time  mean_score_time  std_score_time  \
0       0.022856      0.006778         0.000000        0.000000   
1       0.052126      0.008260         0.007915        0.001868   
2       0.072857      0.002219         0.010156        0.000036   
3       0.494088      0.003667         0.030780        0.000354   
4       0.748828      0.006475         0.044039        0.005006   

  param_n_estimators                 params  split0_test_score  \
0                  5    {'n_estimators': 5}           0.851419   
1                 10   {'n_estimators': 10}           0.894825   
2                 15   {'n_estimators': 15}           0.911519   
3                100  {'n_estimators': 100}           0.934891   
4                150  {'n_estimators': 150}           0.933222   

   split1_test_score  split2_test_score  mean_test_score  std_test_score  \
0           0.856427           0.834725         0.847524        0.009278   
1           0.918197           0.899833         0.904285        0.010048   
2           0.936561           0.901503         0.916528        0.014744   
3           0.951586           0.928214         0.938230        0.009829   
4           0.961603           0.926544         0.940456        0.015199   

   rank_test_score  
0                5  
1                4  
2                3  
3                2  
4                1

In [62]:
df[['param_n_estimators','mean_test_score','rank_test_score']]

param_n_estimators  mean_test_score  rank_test_score
0                  5         0.847524                5
1                 10         0.904285                4
2                 15         0.916528                3
3                100         0.938230                2
4                150         0.940456                1

### GridSearchCV for SVM

In [63]:
svm=GridSearchCV(SVC(),{
    'C':[0.5,1,5],
    'kernel':['rbf','linear','sigmoid','poly'],
    'gamma':['scale','auto']
},return_train_score=False,cv=3
)

In [64]:
svm.fit(digit.data,digit.target)

GridSearchCV(cv=3, estimator=SVC(),
             param_grid={'C': [0.5, 1, 5], 'gamma': ['scale', 'auto'],
                         'kernel': ['rbf', 'linear', 'sigmoid', 'poly']})

In [65]:
svm_df=pd.DataFrame(svm.cv_results_)
svm_df

mean_fit_time  std_fit_time  mean_score_time  std_score_time param_C  \
0        0.117631      0.005160         0.067425        0.003137     0.5   
1        0.053320      0.005174         0.029473        0.000895     0.5   
2        0.269216      0.005767         0.137352        0.003707     0.5   
3        0.069210      0.003503         0.041058        0.001092     0.5   
4        0.458424      0.018354         0.127125        0.001836     0.5   
5        0.058267      0.004194         0.028017        0.000622     0.5   
6        0.347883      0.011155         0.105677        0.006274     0.5   
7        0.059615      0.006911         0.035208        0.002676     0.5   
8        0.101087      0.003170         0.059836        0.004372       1   
9        0.055562      0.004575         0.033905        0.004479       1   
10       0.207230      0.010873         0.116018        0.003261       1   
11       0.060476      0.001982         0.034316        0.003329       1   
12       0.446607      0.008788         0.119155        0.003342       1   
13       0.052524      0.001780         0.026969        0.003260       1   
14       0.340397      0.000782         0.101865        0.001098       1   
15       0.062715      0.002888         0.034380        0.003142       1   
16       0.084639      0.003495         0.050069        0.000967       5   
17       0.052141      0.002003         0.028931        0.004700       5   
18       0.109103      0.002713         0.080266        0.001681       5   
19       0.061037      0.000996         0.033603        0.004898       5   
20       0.440102      0.005198         0.123549        0.001875       5   
21       0.050770      0.006714         0.027698        0.003879       5   
22       0.341140      0.005069         0.100205        0.003876       5   
23       0.060915      0.003991         0.034505        0.005980       5   

   param_gamma param_kernel  \
0        scale          rbf   
1        scale       linear   
2        scale      sigmoid   
3        scale         poly   
4         auto          rbf   
5         auto       linear   
6         auto      sigmoid   
7         auto         poly   
8        scale          rbf   
9        scale       linear   
10       scale      sigmoid   
11       scale         poly   
12        auto          rbf   
13        auto       linear   
14        auto      sigmoid   
15        auto         poly   
16       scale          rbf   
17       scale       linear   
18       scale      sigmoid   
19       scale         poly   
20        auto          rbf   
21        auto       linear   
22        auto      sigmoid   
23        auto         poly   

                                               params  split0_test_score  \
0       {'C': 0.5, 'gamma': 'scale', 'kernel': 'rbf'}           0.958264   
1    {'C': 0.5, 'gamma': 'scale', 'kernel': 'linear'}           0.934891   
2   {'C': 0.5, 'gamma': 'scale', 'kernel': 'sigmoid'}           0.881469   
3      {'C': 0.5, 'gamma': 'scale', 'kernel': 'poly'}           0.958264   
4        {'C': 0.5, 'gamma': 'auto', 'kernel': 'rbf'}           0.135225   
5     {'C': 0.5, 'gamma': 'auto', 'kernel': 'linear'}           0.934891   
6    {'C': 0.5, 'gamma': 'auto', 'kernel': 'sigmoid'}           0.100167   
7       {'C': 0.5, 'gamma': 'auto', 'kernel': 'poly'}           0.953255   
8         {'C': 1, 'gamma': 'scale', 'kernel': 'rbf'}           0.964942   
9      {'C': 1, 'gamma': 'scale', 'kernel': 'linear'}           0.934891   
10    {'C': 1, 'gamma': 'scale', 'kernel': 'sigmoid'}           0.869783   
11       {'C': 1, 'gamma': 'scale', 'kernel': 'poly'}           0.958264   
12         {'C': 1, 'gamma': 'auto', 'kernel': 'rbf'}           0.380634   
13      {'C': 1, 'gamma': 'auto', 'kernel': 'linear'}           0.934891   
14     {'C': 1, 'gamma': 'auto', 'kernel': 'sigmoid'}           0.100167   
15        {'C': 1, 'gamma': 'auto', 'kernel': 'poly'}           0.953255   
16        {'C': 5, 'gamma': 'scale'

In [66]:
svm_df[['param_C','param_gamma','param_kernel','mean_test_score','rank_test_score']]

param_C param_gamma param_kernel  mean_test_score  rank_test_score
0      0.5       scale          rbf         0.957707                9
1      0.5       scale       linear         0.943795               10
2      0.5       scale      sigmoid         0.893155               16
3      0.5       scale         poly         0.966055                3
4      0.5        auto          rbf         0.115192               21
5      0.5        auto       linear         0.943795               10
6      0.5        auto      sigmoid         0.101280               22
7      0.5        auto         poly         0.960490                5
8        1       scale          rbf         0.969950                2
9        1       scale       linear         0.943795               10
10       1       scale      sigmoid         0.884252               17
11       1       scale         poly         0.963272                4
12       1        auto          rbf         0.434613               20
13       1        auto       linear         0.943795               10
14       1        auto      sigmoid         0.101280               22
15       1        auto         poly         0.960490                5
16       5       scale          rbf         0.973289                1
17       5       scale       linear         0.943795               10
18       5       scale      sigmoid         0.815804               18
19       5       scale         poly         0.960490                5
20       5        auto          rbf         0.460211               19
21       5        auto       linear         0.943795               10
22       5        auto      sigmoid         0.101280               22
23       5        auto         poly         0.960490                5

In [67]:
top_5=svm_df[svm_df.rank_test_score<=5][['param_C','param_gamma','param_kernel','mean_test_score','rank_test_score']]
top_5

param_C param_gamma param_kernel  mean_test_score  rank_test_score
3      0.5       scale         poly         0.966055                3
7      0.5        auto         poly         0.960490                5
8        1       scale          rbf         0.969950                2
11       1       scale         poly         0.963272                4
15       1        auto         poly         0.960490                5
16       5       scale          rbf         0.973289                1
19       5       scale         poly         0.960490                5
23       5        auto         poly         0.960490                5

In [68]:
top_5.sort_values(by=['rank_test_score'])

param_C param_gamma param_kernel  mean_test_score  rank_test_score
16       5       scale          rbf         0.973289                1
8        1       scale          rbf         0.969950                2
3      0.5       scale         poly         0.966055                3
11       1       scale         poly         0.963272                4
7      0.5        auto         poly         0.960490                5
15       1        auto         poly         0.960490                5
19       5       scale         poly         0.960490                5
23       5        auto         poly         0.960490                5

In [69]:
svm.best_params_

{'C': 5, 'gamma': 'scale', 'kernel': 'rbf'}

In [70]:
svm.best_score_

0.9732888146911519

### To the find the best performaning model with parameters

In [71]:
models={
    'svm':{
        'model':SVC(),
        'params':{
            'C':[0.5,1,5],
            'gamma':['auto','scale'],
            'kernel':['rbf','linear','poly']
        }
    },
    'random_forest':{
        'model':RandomForestClassifier(),
        'params':{
            'n_estimators':[10,50,100,150,200],
        }
    },
    'logistic':{
        'model':LogisticRegression(),
        'params':{
            'solver':['newton-cg', 'lbfgs', 'liblinear', 'sag', 'saga'],
            'multi_class':['auto', 'ovr', 'multinomial']
        }       
    },
    'gaussian':{
        'model':GaussianNB(),
        'params':{}
    },
    'Multinomial':{
        'model':MultinomialNB(),
        'params':{}
    },
    'decisionTree':{
        'model':DecisionTreeClassifier(),
        'params':{
            'criterion':['gini', 'entropy'],
            'splitter':['best','random']
        }
    }
}

In [72]:
scores=[]
for model_name,mp in models.items():
    cv=GridSearchCV(mp['model'],mp['params'],cv=5,return_train_score=False)
    cv.fit(digit.data,digit.target)
    scores.append({
        'model':model_name,
        'best_params':cv.best_params_,
        'best_score':cv.best_score_
        }
    )

In [73]:
scores

[{'model': 'svm',
  'best_params': {'C': 5, 'gamma': 'scale', 'kernel': 'rbf'},
  'best_score': 0.9738502011761063},
 {'model': 'random_forest',
  'best_params': {'n_estimators': 50},
  'best_score': 0.9426957598266791},
 {'model': 'logistic',
  'best_params': {'multi_class': 'ovr', 'solver': 'sag'},
  'best_score': 0.9237867533271433},
 {'model': 'gaussian', 'best_params': {}, 'best_score': 0.8069281956050759},
 {'model': 'Multinomial', 'best_params': {}, 'best_score': 0.8703497369235531},
 {'model': 'decisionTree',
  'best_params': {'criterion': 'entropy', 'splitter': 'best'},
  'best_score': 0.8119374806561437}]

In [74]:
score=pd.DataFrame(scores)
score

model                                   best_params  best_score
0            svm   {'C': 5, 'gamma': 'scale', 'kernel': 'rbf'}    0.973850
1  random_forest                          {'n_estimators': 50}    0.942696
2       logistic       {'multi_class': 'ovr', 'solver': 'sag'}    0.923787
3       gaussian                                            {}    0.806928
4    Multinomial                                            {}    0.870350
5   decisionTree  {'criterion': 'entropy', 'splitter': 'best'}    0.811937

### Therefore SVC {'C': 5, 'gamma': 'scale', 'kernel': 'rbf'} is the best model with 97.3% score.